File: go/arith.go
/* Copyright (c) 2025 Richard Rodger, MIT License */

package aontu

import (
	"math"
	"math/big"
)


func divides(op string) bool {
	return "div" == op || "mod" == op || "rem" == op
}

func arith(ctx *Ctx, op string, node Val, a, b Val) Val {
	return arithNamed(ctx, op, op, node, a, b)
}

func arithNamed(ctx *Ctx, op, name string, node Val, a, b Val) Val {
	av, aok := unpref(a).(*ScalarVal)
	bv, bok := unpref(b).(*ScalarVal)

	if !aok || !bok || !arithKinds[av.kind] || !arithKinds[bv.kind] {
		return makeNilErrFull(ctx, "invalid-arg", node, nil, name, nil)
	}

	afloat := av.kind == KindFloat
	bfloat := bv.kind == KindFloat

	// A big leaf never silently becomes a binary float, in EITHER
	// operand order. The error names both leaves in operand order.
	if (afloat && isBigKind(bv.kind)) || (isBigKind(av.kind) && bfloat) {
		return makeNilErrFull(ctx, "exact_float_mix", node, nil, name,
			map[string]string{
				"left": av.kind.String(), "right": bv.kind.String()})
	}

	if afloat || bfloat {
		return floatArith(ctx, op, name, node,
			primFloat(av.peg), primFloat(bv.peg))
	}

	if av.kind == KindBigDecimal || bv.kind == KindBigDecimal {
		return decimalArith(ctx, op, name, node,
			scalarDecimal(av), scalarDecimal(bv))
	}

	return integerArith(ctx, op, name, node, scalarBigInt(av), scalarBigInt(bv),
		av.kind == KindBigInteger || bv.kind == KindBigInteger)
}

var arithKinds = map[Kind]bool{
	KindInteger:    true,
	KindFloat:      true,
	KindBigInteger: true,
	KindBigDecimal: true,
}

func isBigKind(k Kind) bool {
	return k == KindBigInteger || k == KindBigDecimal
}

func floatArith(ctx *Ctx, op, name string, node Val, x, y float64) Val {
	if divides(op) && 0 == y {
		return makeNilErrFull(ctx, "divide_by_zero", node, nil, name, nil)
	}

	var out float64
	switch op {
	case "add":
		out = x + y
	case "sub":
		out = x - y
	case "mul":
		out = x * y
	case "div":
		out = x / y
	case "rem":
		// Truncated remainder, sign following the DIVIDEND, which is
		// what math.Mod gives (despite the name) and what JavaScript's
		// `%` gives.
		out = math.Mod(x, y)
	default:
		out = math.Mod(x, y)
		if 0 != out && (out < 0) != (y < 0) {
			out += y
		}
	}

	if math.IsInf(out, 0) || math.IsNaN(out) {
		return makeNilErrFull(ctx, "float_overflow", node, nil, name, nil)
	}
	return newFloat(out)
}

func integerArith(ctx *Ctx, op, name string, node Val, x, y *big.Int, big64 bool) Val {
	if divides(op) && 0 == y.Sign() {
		return makeNilErrFull(ctx, "divide_by_zero", node, nil, name, nil)
	}

	out := new(big.Int)
	switch op {
	case "add":
		out.Add(x, y)
	case "sub":
		out.Sub(x, y)
	case "mul":
		out.Mul(x, y)
	case "div":
		out.Quo(x, y)
	case "rem":
		out.Rem(x, y)
	default:
		out.Rem(x, y)
		if 0 != out.Sign() && (out.Sign() < 0) != (y.Sign() < 0) {
			out.Add(out, y)
		}
	}

	if big64 {
		// Unbounded and exact: nothing to check, and no demotion to
		// integer however small the result.
		return newBigInteger(out)
	}

	if !isIntegerStorable(out) {
		return makeNilErrFull(ctx, "inexact_integer_sum", node, nil, name,
			map[string]string{"sum": out.String()})
	}
	return newInteger(out.Int64())
}

// decimalArith is the decimal leaf. Addition, subtraction and
// multiplication are exact coefficient arithmetic and land here;
// division does not, and says so.
func decimalArith(ctx *Ctx, op, name string, node Val, x, y *Decimal) Val {
	if divides(op) {
		return makeNilErrFull(ctx, "inexact_divide", node, nil, name, nil)
	}

	var out *Decimal
	switch op {
	case "add":
		out = x.add(y)
	case "sub":
		out = x.add(y.neg())
	default:
		out = x.mul(y)
	}

	// The budget applies to RESULTS as well as literals: an exact answer
	// too wide to hold is refused, never rounded to fit.
	if out.overBudget() {
		return makeNilErrFull(ctx, "decimal_budget", node, nil, name, nil)
	}
	return newBigDecimal(out)
}

File: ts/src/val/arith.ts
/* Copyright (c) 2025 Richard Rodger, MIT License */


import type { Val } from '../type'

import { AontuContext } from '../ctx'
import { makeNilErr } from '../err'

import { IntegerVal } from './IntegerVal'
import { NumberVal } from './NumberVal'
import { BigIntegerVal } from './BigIntegerVal'
import { BigDecimalVal } from './BigDecimalVal'
import { Decimal, decimalOverBudget } from './Decimal'
import { isIntegerStorable } from './numkind'


type ArithOp = 'add' | 'sub' | 'mul' | 'div' | 'mod' | 'rem'

// The three that divide, and therefore the three that can be handed a
// zero divisor and cannot answer over the decimal leaf.
function divides(op: ArithOp): boolean {
  return 'div' === op || 'mod' === op || 'rem' === op
}


// The numeric leaves, told apart by their own flags rather than by the
// JavaScript type of the peg: `integer` and `float` share `number`.
type ArithKind = 'integer' | 'float' | 'biginteger' | 'bigdecimal'

const EXACT_RANK: Record<string, number> = {
  integer: 1,
  biginteger: 2,
  bigdecimal: 3,
}


function isBig(k: ArithKind): boolean {
  return 'biginteger' === k || 'bigdecimal' === k
}


// A pref operand contributes its preferred value, and therefore that
// value's kind too -- the same rule `+` applies.
function unpref(v: any): any {
  while (v?.isPref) {
    v = v.peg
  }
  return v
}


function arithKind(v: any): ArithKind | undefined {
  if (true !== v?.isScalar) {
    return undefined
  }
  if (v.isBigInteger) {
    return 'biginteger'
  }
  if (v.isBigDecimal) {
    return 'bigdecimal'
  }
  if (v.isInteger) {
    return 'integer'
  }
  return 'number' === typeof v.peg ? 'float' : undefined
}


function asInteger(v: any, k: ArithKind): bigint {
  return 'biginteger' === k ? v.peg : BigInt(v.peg)
}


function asDecimal(v: any, k: ArithKind): Decimal {
  return 'bigdecimal' === k ? v.peg : new Decimal(asInteger(v, k), 0)
}


function arith(
  ctx: AontuContext | undefined,
  op: ArithOp,
  node: Val,
  a: Val,
  b: Val,
  attempt?: string
): Val {
  const name = attempt ?? op
  const av: any = unpref(a)
  const bv: any = unpref(b)
  const ak = arithKind(av)
  const bk = arithKind(bv)

  if (undefined === ak || undefined === bk) {
    return makeNilErr(ctx, 'invalid-arg', node, undefined, name)
  }

  // A big leaf never silently becomes a binary float, in EITHER operand
  // order. The error names both leaves in operand order.
  if (('float' === ak && isBig(bk)) || (isBig(ak) && 'float' === bk)) {
    return makeNilErr(ctx, 'exact_float_mix', node, undefined, name,
      { left: ak, right: bk })
  }

  if ('float' === ak || 'float' === bk) {
    return floatArith(ctx, op, name, node, av.peg, bv.peg)
  }

  const rank = EXACT_RANK[bk] < EXACT_RANK[ak] ? EXACT_RANK[ak] : EXACT_RANK[bk]

  if (EXACT_RANK.bigdecimal === rank) {
    return decimalArith(ctx, op, name, node, asDecimal(av, ak), asDecimal(bv, bk))
  }

  return integerArith(ctx, op, name, node, asInteger(av, ak),
    asInteger(bv, bk), EXACT_RANK.biginteger === rank)
}


function floatArith(
  ctx: AontuContext | undefined,
  op: ArithOp,
  name: string,
  node: Val,
  x: number,
  y: number
): Val {
  if (divides(op) && 0 === y) {
    return makeNilErr(ctx, 'divide_by_zero', node, undefined, name)
  }

  const out =
    'add' === op ? x + y :
      'sub' === op ? x - y :
        'mul' === op ? x * y :
          'div' === op ? x / y :
            // Truncated remainder, sign following the DIVIDEND, which is
            // what JavaScript's `%` and Go's math.Mod both give...
            'rem' === op ? x % y :
              flooredMod(x % y, y)

  return Number.isFinite(out) ?
    new NumberVal({ peg: out }) :
    makeNilErr(ctx, 'float_overflow', node, undefined, name)
}


function flooredMod(rem: number, y: number): number {
  return 0 !== rem && (rem < 0) !== (y < 0) ? rem + y : rem
}


// The exact integral leaves. Both compute in bigint, so nothing passes
// through binary64 and nothing rounds; only the storage test at the end
// differs, because `biginteger` is unbounded and `integer` is not.
function integerArith(
  ctx: AontuContext | undefined,
  op: ArithOp,
  name: string,
  node: Val,
  x: bigint,
  y: bigint,
  big: boolean
): Val {
  if (divides(op) && 0n === y) {
    return makeNilErr(ctx, 'divide_by_zero', node, undefined, name)
  }

  const out =
    'add' === op ? x + y :
      'sub' === op ? x - y :
        'mul' === op ? x * y :
          'div' === op ? x / y :
            'rem' === op ? x % y :
              flooredModBig(x % y, y)

  if (big) {
    // Unbounded and exact: nothing to check, and no demotion to
    // `integer` however small the result.
    return new BigIntegerVal({ peg: out })
  }

  return isIntegerStorable(out) ?
    new IntegerVal({ peg: Number(out) }) :
    makeNilErr(ctx, 'inexact_integer_sum', node, undefined, name,
      { sum: out.toString() })
}


function flooredModBig(rem: bigint, y: bigint): bigint {
  return 0n !== rem && (rem < 0n) !== (y < 0n) ? rem + y : rem
}


// The decimal leaf. Addition, subtraction and multiplication are exact
// coefficient arithmetic and land here; division does not, and says so.
function decimalArith(
  ctx: AontuContext | undefined,
  op: ArithOp,
  name: string,
  node: Val,
  x: Decimal,
  y: Decimal
): Val {
  if (divides(op)) {
    return makeNilErr(ctx, 'inexact_divide', node, undefined, name)
  }

  const out =
    'add' === op ? x.add(y) :
      'sub' === op ? x.add(y.negate()) :
        x.multiply(y)

  // The budget applies to RESULTS as well as literals: an exact answer
  // too wide to hold is refused, never rounded to fit.
  return decimalOverBudget(out) ?
    makeNilErr(ctx, 'decimal_budget', node, undefined, name) :
    new BigDecimalVal({ peg: out })
} /* node:coverage ignore next 9 */


export type {
  ArithOp,
}

export {
  arith,
}
